src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/VisualStudioRazorParser.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Timers;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using ITextBuffer = Microsoft.VisualStudio.Text.ITextBuffer;
+using Timer = System.Timers.Timer;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Editor
+{
+    public class VisualStudioRazorParser : ForegroundThreadAffinitizedObject, IDisposable
+    {
+        // Internal for testing.
+        internal readonly ITextBuffer _textBuffer;
+        internal readonly Timer _idleTimer;
+
+        private const int IdleDelay = 3000;
+        private readonly ICompletionBroker _completionBroker;
+        private readonly BackgroundParser _parser;
+        private RazorSyntaxTreePartialParser _partialParser;
+
+        public VisualStudioRazorParser(ITextBuffer buffer, RazorTemplateEngine templateEngine, string filePath, ICompletionBroker completionBroker)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (templateEngine == null)
+            {
+                throw new ArgumentNullException(nameof(templateEngine));
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            if (completionBroker == null)
+            {
+                throw new ArgumentNullException(nameof(completionBroker));
+            }
+
+            TemplateEngine = templateEngine;
+            FilePath = filePath;
+            _textBuffer = buffer;
+            _completionBroker = completionBroker;
+            _textBuffer.Changed += TextBuffer_OnChanged;
+            _parser = new BackgroundParser(templateEngine, filePath);
+            _idleTimer = new Timer(IdleDelay);
+            _idleTimer.Elapsed += Onidle;
+            _parser.ResultsReady += OnResultsReady;
+
+            _parser.Start();
+        }
+
+        public event EventHandler<DocumentStructureChangedEventArgs> DocumentStructureChanged;
+
+        public RazorTemplateEngine TemplateEngine { get; }
```

Comment:
This and FilePath don't technically need to be public once we control the lifetime of `VisualStudioRazorParser`.